src/vm/arm/asmhelpers.S

```diff
@@ -0,0 +1,1358 @@
+        ldr     r3, [r0, #(MachState___R4_R11 + ((\regIndex-4)*4))]
+
+        cmp     r2, r3
+        bne     0f
+        ldr     \reg, [r2]
+0:
+
+        .endm
+
+// EXTERN_C int __fastcall HelperMethodFrameRestoreState(
+//         INDEBUG_COMMA(HelperMethodFrame *pFrame)
+//         MachState *pState
+//         )
+        LEAF_ENTRY HelperMethodFrameRestoreState, _TEXT
+
+#ifdef _DEBUG
+        mov r0, r1
+#endif
+
+        // If machine state is invalid, then simply exit
+        ldr r1, [r0, #MachState__isValid]
+        cmp r1, #0
+        beq Done
+
+        RestoreRegMS 4, R4
+        RestoreRegMS 5, R5
+        RestoreRegMS 6, R6
+        RestoreRegMS 7, R7
+        RestoreRegMS 8, R8
+        RestoreRegMS 9, R9
+        RestoreRegMS 10, R10
+        RestoreRegMS 11, R11
+Done:
+        // Its imperative that the return value of HelperMethodFrameRestoreState is zero
+        // as it is used in the state machine to loop until it becomes zero.
+        // Refer to HELPER_METHOD_FRAME_END macro for details.
+        mov r0,#0
+        bx lr
+
+        LEAF_END HelperMethodFrameRestoreState, _TEXT
+
+#if 0
+// ------------------------------------------------------------------
+// Macro to generate Redirection Stubs
+//
+// $reason : reason for redirection
+//                     Eg. GCThreadControl
+// NOTE: If you edit this macro, make sure you update GetCONTEXTFromRedirectedStubStackFrame.
+// This function is used by both the personality routine and the debugger to retrieve the original CONTEXT.
+        .macro GenerateRedirectedHandledJITCaseStub reason
+
+        NESTED_ENTRY RedirectedHandledJITCaseFor\reason\()_Stub, _TEXT, NoHandler
+
+        push {r7,lr}     // return address
+        alloc_stack 4    // stack slot to save the CONTEXT *
+        mov r7, sp
+
+        //REDIRECTSTUB_SP_OFFSET_CONTEXT is defined in asmconstants.h
+        //If CONTEXT is not saved at 0 offset from SP it must be changed as well.
+        //ASSERT REDIRECTSTUB_SP_OFFSET_CONTEXT == 0
+
+        // Runtime check for 8-byte alignment. This check is necessary as this function can be
+        // entered before complete execution of the prolog of another function.
+        and r0, r7, #4
+        sub sp, sp, r0
+
+        // stack must be 8 byte aligned
+        //CHECK_STACK_ALIGNMENT
+
+        //
+        // Save a copy of the redirect CONTEXT*.
+        // This is needed for the debugger to unwind the stack.
+        //
+        bl GetCurrentSavedRedirectContext
+        str r0, [r7]
+
+        //
+        // Fetch the interrupted pc and save it as our return address.
+        //
+        ldr r1, [r0, #CONTEXT_Pc]
+        str r1, [r7, #8]
+
+        //
+        // Call target, which will do whatever we needed to do in the context
+        // of the target thread, and will RtlRestoreContext when it is done.
+        //
+        bl _RedirectedHandledJITCaseFor\reason\()_Stub@Thread@@CAXXZ
+
+        //EMIT_BREAKPOINT // Unreachable
```

Comment:
Ah yes, I can implement EMIT_BREAKPOINT and CHECK_STACK_ALIGNMENT in the ARM specific file now. Both appeared to be debugging functions so I didn't implement them immediately.